usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1077,25 +1082,30 @@ zil_lwb_commit(zilog_t *zilog, itx_t *itx, lwb_t *lwb)
 
 	zil_lwb_write_init(zilog, lwb);
 
+cont:
 	/*
 	 * If this record won't fit in the current log block, start a new one.
+	 * For WR_NEED_COPY optimize layout for minimal number of chunks, but
+	 * try to keep wasted space withing reasonable range (12%).
 	 */
-	if (lwb->lwb_nused + reclen + dlen > lwb->lwb_sz) {
+	lwb_sp = lwb->lwb_sz - lwb->lwb_nused;
+	if (reclen > lwb_sp || (reclen + dlen > lwb_sp &&
```

Comment:
Even with the comment above, the purpose of these additional checks in this conditional still isn't clear to me. Could you try to add some more info to describe why we want these additional checks (i.e. this line, and the two below)?

I'm still a little new to the details of the ZIL, so perhaps my inexperience is showing.